### PR TITLE
add test for views.py

### PR DESCRIPTION
**Pull request by speedy93**
image

image

https://kindergarten.stxnext.pl/browse/WDP-115

### CODE REVIEW

File: src/presence_analyzer/utils.py [hunk 1]
```diff
@@ -7,12 +7,12 @@
 from json import dumps
 from functools import wraps
 from datetime import datetime
+import logging
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: https://www.python.org/dev/peps/pep-0008/#imports

File: src/presence_analyzer/views.py [hunk 2]
```diff
@@ -43,6 +45,7 @@ def mean_time_weekday_view(user_id):
     data = get_data()
     if user_id not in data:
         log.debug('User %s not found!', user_id)
+        # print log.debug
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: ?

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -53,6 +54,32 @@ def test_api_users(self):
...
+        """
+        Test by user id.
+        """
+        resp = self.client.get('/api/v1/presence_weekday/0')
+        self.assertEqual(resp.status_code, 404)
+        resp = self.client.get('/api/v1/presence_weekday/11')
+        self.assertEqual(resp.status_code, 200)
+        expected = '[["Weekday", "Presence (s)"], ["Mon", 24123], ' +\
+            '["Tue", 16564], ["Wed", 25321], ["Thu", 45968], ' +\
+            '["Fri", 6426], ["Sat", 0], ["Sun", 0]]'
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Style https://gist.github.com/krotkiewicz/3134c608c91cddbf21e5
Dont compare jsons, use json.loads and compare dicts

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -53,6 +54,32 @@ def test_api_users(self):
...
+            '["Wed", 25321.0], ["Thu", 22984.0], ' +\
+            '["Fri", 6426.0], ["Sat", 0], ["Sun", 0]]'
+        self.assertEqual(resp.data, expected)
+
+    def test_presence_weekday_view(self):
+        """
+        Test by user id.
+        """
+        resp = self.client.get('/api/v1/presence_weekday/0')
+        self.assertEqual(resp.status_code, 404)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: enter

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -53,6 +54,32 @@ def test_api_users(self):
...
+        resp = self.client.get('/api/v1/presence_weekday/11')
+        self.assertEqual(resp.status_code, 200)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: enter

File: runtime/data/test_data.csv [hunk 6]
```diff
@@ -6,4 +6,4 @@
 11,2013-09-10,09:19:50,13:55:54
 11,2013-09-11,09:13:26,16:15:27
 11,2013-09-12,10:18:36,16:41:25
-11,2013-09-13,13:16:56,15:04:02
\ No newline at end of file
+11,2013-09-13,13:16:56,15:04:02
```
Comment (piotrgrundas) [inline runtime/data/test_data.csv]: revert

File: src/presence_analyzer/main.py [hunk 7]
```diff
@@ -13,6 +13,6 @@
 
 app = Flask(__name__)  # pylint: disable=invalid-name
 app.config.update(
-    DEBUG = True,
-    DATA_CSV = MAIN_DATA_CSV
+    DEBUG=True,
+    DATA_CSV=MAIN_DATA_CSV
```
Comment (piotrgrundas) [inline src/presence_analyzer/main.py]: This shouldn't be here

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -86,6 +113,15 @@ def test_get_data(self):
...
 
+    def test_group_by_weekday(self):
+        """
+        Test list of times of user.
+        """
+        data = utils.get_data()
+        actual = utils.group_by_weekday(data[10])
+        expected = [[], [30047], [24465], [23705], [], [], []]
+        self.assertEqual(actual, expected)
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: And where are tests for rest utils functions?
```
seconds_since_midnight
interval
mean
```

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -86,6 +113,15 @@ def test_get_data(self):
             datetime.time(9, 39, 5)
         )
 
+    def test_group_by_weekday(self):
+        """
+        Test list of times of user.
+        """
+        data = utils.get_data()
+        actual = utils.group_by_weekday(data[10])
+        expected = [[], [30047], [24465], [23705], [], [], []]
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: enter

File: src/presence_analyzer/utils.py (hunk 1 above)
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: move to top

File: src/presence_analyzer/tests.py [hunk 10]
```diff
@@ -39,20 +41,71 @@ def test_mainpage(self):
...
 
+    def test_mean_time_weekday_view(self):
+        """
+        Test by user id.
+        """
+        resp = self.client.get('/api/v1/mean_time_weekday/0')
+        self.assertEqual(resp.status_code, 404)
+
+        resp = self.client.get('/api/v1/mean_time_weekday/11')
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: remove

File: src/presence_analyzer/tests.py [hunk 11]
```diff
@@ -39,20 +41,71 @@ def test_mainpage(self):
...
+        self.assertEqual(resp.status_code, 200)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: move before `self.assertDictEqual(data, expected)`

File: src/presence_analyzer/tests.py [hunk 12]
```diff
@@ -39,20 +41,71 @@ def test_mainpage(self):
...
 
     def test_api_users(self):
         """
         Test users listing.
         """
         resp = self.client.get('/api/v1/users')
+
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]:
```
resp
data

asserts
```

File: src/presence_analyzer/tests.py [hunk 13]
```diff
@@ -86,6 +139,59 @@ def test_get_data(self):
...
+        )
+        time_two = datetime.timedelta(
+            hours=11,
+            minutes=1,
+            seconds=36,
+        )
+        actual = utils.interval(start, end)
+        expected = time_one - time_two
+
+        self.assertEqual(expected.seconds, abs(actual))
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: remove time_one, time_two. In tests expected results are usually just hardcoded, u don't even need to create `expected` variable, just pass it directly to assert

File: src/presence_analyzer/tests.py [hunk 14]
```diff
@@ -86,6 +139,59 @@ def test_get_data(self):
...
+
+    def test_mean(self):
+        """
+        Test arithmetic mean.
+        """
+        items = [1, 2, 3, 4, 5, 6]
+        actual = utils.mean(items)
+        self.assertEqual(actual, 3.5)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Provide more test cases to all tests, not only 1 case, test some edge cases. 
And in such simple test u could simply do

`self.assertEqual(utils.mean([1, 2, 3 ...]), 3.5)`

File: src/presence_analyzer/tests.py [hunk 15]
```diff
@@ -86,6 +137,51 @@ def test_get_data(self):
...
+    def test_group_by_weekday(self):
+        """
+        Test list of times of user.
+        """
+        data = utils.get_data()
+        actual = utils.group_by_weekday(data[10])
+        expected = [[], [30047], [24465], [23705], [], [], []]
+
+        self.assertEqual(actual, expected)
+        self.assertIsInstance(actual, list)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: you check it already in assert before

File: src/presence_analyzer/tests.py [hunk 16]
```diff
@@ -86,6 +137,51 @@ def test_get_data(self):
             datetime.time(9, 39, 5)
         )
 
+    def test_group_by_weekday(self):
+        """
+        Test list of times of user.
+        """
+        data = utils.get_data()
+        actual = utils.group_by_weekday(data[10])
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Why u assign something that u use only once?

File: src/presence_analyzer/tests.py [hunk 17]
```diff
@@ -86,6 +137,51 @@ def test_get_data(self):
...
+        self.assertEqual(25300, abs(actual))
+        self.assertIsInstance(actual, int)
+
+    def test_mean(self):
+        """
+        Test arithmetic mean.
+        """
+        self.assertEqual(utils.mean([1, 2, 3, 4, 5, 6]), 3.5)
+        self.assertEqual(utils.mean([]), 0)
+        self.assertIsInstance(utils.mean([]), int)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 18]
```diff
@@ -86,6 +137,51 @@ def test_get_data(self):
...
+    def test_interval(self):
+        """
+        Test inverval in seconds between two datetime.
+        """
+        start = datetime.time(18, 3, 16)
+        end = datetime.time(11, 1, 36)
+        actual = utils.interval(start, end)
+
+        self.assertEqual(25300, abs(actual))
+        self.assertIsInstance(actual, int)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 19]
```diff
@@ -86,6 +137,51 @@ def test_get_data(self):
...
+        self.assertEqual(expected_time.seconds, data)
+        self.assertIsInstance(data, int)
+
+    def test_interval(self):
+        """
+        Test inverval in seconds between two datetime.
+        """
+        start = datetime.time(18, 3, 16)
+        end = datetime.time(11, 1, 36)
+        actual = utils.interval(start, end)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: again assign only for using it once?

File: src/presence_analyzer/tests.py [hunk 20]
```diff
@@ -86,6 +137,51 @@ def test_get_data(self):
             datetime.time(9, 39, 5)
         )
 
+    def test_group_by_weekday(self):
+        """
+        Test list of times of user.
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: More info please

File: src/presence_analyzer/tests.py [hunk 21]
```diff
@@ -39,20 +41,68 @@ def test_mainpage(self):
...
+                'Thu': 22984.0,
+                'Fri': 6426.0,
+                'Sat': 0,
+                'Sun': 0,
+            }
+        )
+
+    def test_presence_weekday_view(self):
+        """
+        Test by user id.
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: More info please

File: src/presence_analyzer/views.py [hunk 22]
```diff
@@ -4,12 +4,14 @@
...
 
 import calendar
+import logging
+
 from flask import redirect, abort
 
-from presence_analyzer.main import app
-from presence_analyzer.utils import jsonify, get_data, mean, group_by_weekday
+from main import app  # pylint: disable=relative-import
+from utils import jsonify, get_data, mean, group_by_weekday
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: sort imports

File: src/presence_analyzer/tests.py [hunk 23]
```diff
@@ -86,6 +141,69 @@ def test_get_data(self):
...
+                         datetime.timedelta(
+                             hours=12,
+                             minutes=0,
+                             seconds=35).seconds)
+        self.assertEqual(utils.seconds_since_midnight(
+            datetime.time(1, 1, 59)),
+                         datetime.timedelta(
+                             hours=1,
+                             minutes=1,
+                             seconds=59).seconds)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: self.assertEqual(
        utils.seconds_since_midnight(datetime.time(1, 1, 59)),
        datetime.timedelta(
            hours=1,
            minutes=1,
            seconds=59
        ).seconds
    )

File: src/presence_analyzer/tests.py [hunk 24]
```diff
@@ -86,6 +141,69 @@ def test_get_data(self):
...
+            datetime.time(11, 1, 36),
+        )))
+        self.assertEqual(61300, abs(utils.interval(
+            datetime.time(18, 3, 16),
+            datetime.time(1, 1, 36)
+        )))
+        self.assertEqual(3696, abs(utils.interval(
+            datetime.time(0, 0, 0),
+            datetime.time(1, 1, 36),
+        )))
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: self.assertEqual(
        3696, 
        abs(utils.interval(datetime.time(0, 0, 0), datetime.time(1, 1, 36)))
    )

I would import `from datetime import time itp` for convinience since ur using it so often.

File: src/presence_analyzer/utils.py [hunk 25]
```diff
@@ -4,15 +4,17 @@
 """
 
 import csv
+import logging
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: Why? All are bultin. Sort them

File: src/presence_analyzer/utils.py [hunk 26]
```diff
@@ -4,15 +4,17 @@
...
 from json import dumps
 from functools import wraps
 from datetime import datetime
 
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: 1 blank line between different type of imports

File: src/presence_analyzer/tests.py [hunk 27]
```diff
@@ -4,10 +4,12 @@
 """
 import os.path
 import json
-import datetime
 import unittest
+from datetime import date, timedelta, time
 
-from presence_analyzer import main, views, utils
+import views# pylint: disable=relative-import
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: image to spaces before comment like in the reset

File: src/presence_analyzer/tests.py [hunk 28]
```diff
@@ -39,20 +41,73 @@ def test_mainpage(self):
...
+                'Fri': 6426.0,
+                'Sat': 0,
+                'Sun': 0,
+            }
+        )
+
+    def test_presence_weekday_view(self):
+        """
+        Test by user id, for existing user and don't existing.
+        Check correct output when user exist.
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Remove this line

File: src/presence_analyzer/tests.py [hunk 29]
```diff
@@ -39,20 +41,73 @@ def test_mainpage(self):
...
+                'Thu': 22984.0,
+                'Fri': 6426.0,
+                'Sat': 0,
+                'Sun': 0,
+            }
+        )
+
+    def test_presence_weekday_view(self):
+        """
+        Test by user id, for existing user and don't existing.
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Test by user id, for existing and non existing user.

File: src/presence_analyzer/tests.py [hunk 30]
```diff
@@ -76,16 +131,84 @@ def test_get_data(self):
         Test parsing of CSV file.
         """
         data = utils.get_data()
+
         self.assertIsInstance(data, dict)
         self.assertItemsEqual(data.keys(), [10, 11])
-        sample_date = datetime.date(2013, 9, 10)
+        sample_date = date(2013, 9, 10)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]:
```
variables

asserts
```

File: src/presence_analyzer/tests.py [hunk 31]
```diff
@@ -76,16 +131,84 @@ def test_get_data(self):
...
         self.assertEqual(
             data[10][sample_date]['start'],
-            datetime.time(9, 39, 5)
+            time(9, 39, 5)
         )
 
+    def test_group_by_weekday(self):
+        """
+        Test list of times of user.
+        Test output for example value.
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py (hunk 20 above)
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: This doesn't really make sense. What are u actually testing here? What this function do.

File: src/presence_analyzer/tests.py [hunk 32]
```diff
@@ -76,16 +131,84 @@ def test_get_data(self):
...
+                hours=1,
+                minutes=1,
+                seconds=59).seconds
+        )
+
+    def test_interval(self):
+        """
+        Test inverval in seconds between two datetime.
+        """
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: ?

File: src/presence_analyzer/tests.py [hunk 33]
```diff
@@ -76,16 +131,84 @@ def test_get_data(self):
...
+                hours=12,
+                minutes=0,
+                seconds=35).seconds
+        )
+        self.assertEqual(
+            utils.seconds_since_midnight(time(1, 1, 59)),
+            timedelta(
+                hours=1,
+                minutes=1,
+                seconds=59).seconds
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: ??????????

File: src/presence_analyzer/tests.py [hunk 34]
```diff
@@ -76,16 +130,88 @@ def test_get_data(self):
...
+        """
+
+        self.assertEqual(
+            25300,
+            abs(utils.interval(time(18, 3, 16), time(11, 1, 36),))
+        )
+        self.assertEqual(
+            61300,
+            abs(utils.interval(time(18, 3, 16), time(1, 1, 36)))
+        )
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: image
???

File: src/presence_analyzer/tests.py [hunk 35]
```diff
@@ -76,16 +130,88 @@ def test_get_data(self):
         Test parsing of CSV file.
         """
         data = utils.get_data()
+
         self.assertIsInstance(data, dict)
         self.assertItemsEqual(data.keys(), [10, 11])
-        sample_date = datetime.date(2013, 9, 10)
+
+        sample_date = date(2013, 9, 10)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Right after data and group all asserts in this particular case.

File: src/presence_analyzer/tests.py [hunk 36]
```diff
@@ -76,16 +130,88 @@ def test_get_data(self):
...
+            timedelta(
+                hours=1,
+                minutes=1,
+                seconds=59
+            ).seconds
+        )
+
+    def test_interval(self):
+        """
+        Test time in seconds between two times.
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: `Test calculating interval between two time objects.`
